js/runner.go

```diff
@@ -201,6 +202,15 @@ func (r *Runner) newVU(idLocal, idGlobal uint64, samplesOut chan<- stats.SampleC
 		NameToCertificate:  nameToCert,
 		Renegotiation:      tls.RenegotiateFreelyAsClient,
 	}
+	if keyWriter := r.Bundle.RuntimeOptions.KeyWriter; keyWriter.Valid {
+		f, err := os.OpenFile(keyWriter.String, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0600)
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
```

Comment:
This will close the file before anything has had time to write to it. 

We either need to close it at some other time (there is currently no VU cleanup state, and I would argue that will be kind of hard to implement), or we can just not close it with a comment to actually figure out how to close it later on. Arguably in the fast majority of cases this file will stay open until just before k6 gets killed so :shrug:

Reply:
@MStoykov @chtnnh  I think we could open the file from the `run` command then inject it in the `js.Runner` populating a field in the case the option is available. 

```go
type Runner struct {
    NSSKeyLogWriter io.Writer
....
}
```
It should help also for creating a unit test.

Reply:
While it will be nice to open this once and use it throughout, I think doing it in the `run` command will require way too many `lib.Runner` (the interface) changes that I am not particularly certain about and will just be one more thing this interface does (and it already is pretty huge IMO).

But this(the opening of the file) can definitely be moved to the creation of the js.Runner where we first get the RuntimeOptions and keep it as a field. But at least currently, looking at the code around all of this, there is no particular moment the `js.Runner` has no `Stop` method or some way to emulate it that will make the rest of k6 to wait for it to close the file after it is no longer necessary.

Alternatively, we have a context when we activate a VU and a [known place](https://github.com/grafana/k6/blob/b8164eed5413732e94f0f334b9a873146bd0f58b/js/runner.go#L658) where it(vu) no longer runs, but we haven't yet returned it, so the rest of the system will wait for us to ... for example close the file. Opening and closing a file on each activation though might not be great :shrug:. Also this means that we are using 1 file descriptor per VU just for that ... again not great.

We can probably combo the two, so that each time a VU is activated it will create the file and then use it and if it's the last one to be deactivated it will close it :shrug: , but IMO this goes in the same (if not bigger) rabbit hole as the adding of more methods to the `lib.Runner` interface and IMO is completely out of scope. 

I am also not certain that just because we are not running code, some (retry) logic in net/http (or something else as this should be used by all tls connections) will not just try to make a new connection and try to write to an already closed key log file.

So I am mostly for this to be moved in the `New*` methods returning the runner and having it in one place and just having a TODO+issue that we need to close this file. From my experiments, it works great even without closing it and the 1 more open file forever is probably the least problem around the runner ;) 

WDYT @na-- @imiric

Reply:
@MStoykov @na-- 
To echo my comment from my (duplicate as it turns out) solution: I think you may be overthinking this quite a bit :) just think of a use case: it is **only to debug** the comm with a few users at most - never to run at full load, because doing so will compromise the test infra. A few extra file handles on K6' side is nothing compared to the CAP portion (inspect streams, store/dump session) - IMO the only appropriate way to implement large scale capture, is by using an appropriately sized squid/nginx reverse proxy cluster with logging and post-processing.